### ISSUE TITLE
Wrap all `tokio::io::AsyncWrite` types to a specific type which is implemented `AsyncFileWriter`

### ISSUE THREAD

**Issue by ethe** [apache/arrow-rs#6327]
**Is your feature request related to a problem or challenge? Please describe what you are trying to do.**
In current, parquet defines all `AsyncWrite` implement `AsyncFileWriter`:
```rust
impl<T: AsyncWrite + Unpin + Send> AsyncFileWriter for T {
    ... ...
}
```

which would in conflict with:
```rust
impl<T: AsyncFileWriter> AsyncFileWriter for &mut T {
    ... ...
}
```

this definition is not included right now, however, it should be helpful in more general cases.

**Describe the solution you'd like**
Wrap all tokio::io::AsyncWrite into a proxy struct, then we can define `&mut AsyncFileWriter` implements `AsyncFileWriter`:
```rust
pub struct TokioFileWriter<F: AsyncWrite> {
    file: F,
}

impl<F: AsyncWrite> From<F> for TokioFileWriter {
    ... ...
}

impl<F: AsyncWrite + Unpin + Send> for TokioFileWriter {
    ... ...
}

```

**Describe alternatives you've considered**
I don't think there is an alternatives.

**Additional context**
This is a breaking change, but it values to be considered.

**Comment by Xuanwo** [comment]
Hi, could you share the use case for `&mut AsyncFileWriter`?

**Comment by ethe** [comment]
But AsyncArrowWriter is created from AsyncFileWriter, creating an `AsyncArrowWriter` from `&mut AsyncFileWriter` is exactly what I want.

```rust
struct CustomFile {
    // ...
}

impl AsyncFileWriter for CustomFile {
    fn write(
        &mut self,
        bs: bytes::Bytes,
    ) -> futures::future::BoxFuture<'_, parquet::errors::Result<()>> {
        todo!()
    }

    fn complete(&mut self) -> futures::future::BoxFuture<'_, parquet::errors::Result<()>> {
        todo!()
    }
}

fn test() {
    let mut file = CustomFile {};
    {
        let mut writer =
            AsyncArrowWriter::try_new(&mut file, SchemaRef::from(Schema::empty()), None).unwrap();
        // write record batch
        writer.close()
    }
    // do something with file after writes
}
```

**Comment by ethe** [comment]
Also it is helpful to make boxed `AsyncFileWriter` general, compare with `impl AsyncFileWriter for Box<dyn AsyncFileWriter>` currently:
```rust
impl<T: AsyncFileWriter + ?Sized + Unpin + Send> AsyncFileWriter for Box<T> {
    fn write(&mut self, bs: Bytes) -> BoxFuture<'_, Result<()>> {
        self.as_mut().write(bs)
    }

    fn complete(&mut self) -> BoxFuture<'_, Result<()>> {
        self.as_mut().complete()
    }
}
```

**Comment by Xuanwo** [comment]
> // do something with file after writes

From my current understanding, you should not do anything to this file after closing it. Can you provide a detailed list of actions you plan to take?

**Comment by ethe** [comment]
It is my mistake, actually I just want to flush the `AsyncArrowWriter`, then seek and read this file as native file rather than a `AsyncFileWriter`, I have fixed the example code.

**Comment by Xuanwo** [comment]
Understood. I believe you need `AsyncArrowWriter::into_inner`.

---

First of all, allowing users to manipulate the writer after `AsyncArrowWriter` has written the header and data could be error-prone.

Additionally, it doesn't make sense given our context could involve S3, Azure Blob, or GCS storage services. In these contexts, the file doesn't exist before `close` has been called, so you can't perform `seek`/`read` or other actions.

If you simply want to use this API in the local filesystem context, would you consider using [`try_clone`](https://docs.rs/tokio/latest/tokio/fs/struct.File.html#method.try_clone) instead?

**Comment by ethe** [comment]
> Additionally, it doesn't make sense given our context could involve S3, Azure Blob, or GCS storage services. 

It does make sense in local filesystems, which allows to implements `AsyncFileWriter`.

> First of all, allowing users to manipulate the writer after AsyncArrowWriter has written the header and data could be error-prone

However, `&mut tokio::io::AsyncWriter` is a valid `AsyncFileWriter` in the current, if you think parquet should not allow `&mut AsyncFileWriter` be an `AsyncFileWriter`, then I think `&mut tokio::io::AsyncWriter` should be fixed. What ever happends, the behavior of `tokio::io::AsyncWriter` and non-AsyncWriter should be aligned.

**Comment by Xuanwo** [comment]
> It does make sense in local filesystems, which allows to be a `AsyncFileWriter`, don't give up eating for fear of choking.

I agree with you that it useful in the local fs context. Therefore, I propose a new API called `AsyncArrowWriter::into_inner`. I believe this can address your needs in a compatible manner.

**Comment by ethe** [comment]
> Understood. I believe you need AsyncArrowWriter::into_inner.

I think `into_inner` has different behaviors and takes different overheads, it would move the `AsyncFileWriter` but `&mut AsyncFileWriter` is in-placed. But I think it should be nice to have `AsyncArrowWriter::into_inner` method.
